pkg/property/property.go

```diff
@@ -0,0 +1,315 @@
+	Value json.RawMessage `json:"value"`
+}
+
+func (p Property) Validate() error {
+	if len(p.Type) == 0 {
+		return errors.New("type must be set")
+	}
+	if len(p.Value) == 0 {
+		return errors.New("value must be set")
+	}
+	var raw json.RawMessage
+	if err := json.Unmarshal(p.Value, &raw); err != nil {
+		return fmt.Errorf("value is not valid json: %v", err)
+	}
+	return nil
+}
+
+type Package struct {
+	PackageName string `json:"packageName"`
+	Version     string `json:"version"`
+}
+
+type PackageRequired struct {
+	PackageName  string `json:"packageName"`
+	VersionRange string `json:"versionRange"`
+}
+
+type Channel struct {
+	Name     string `json:"name"`
+	Replaces string `json:"replaces,omitempty"`
+}
+
+type GVK struct {
+	Group   string `json:"group"`
+	Kind    string `json:"kind"`
+	Version string `json:"version"`
+}
+
+type GVKRequired struct {
+	Group   string `json:"group"`
+	Kind    string `json:"kind"`
+	Version string `json:"version"`
+}
+
+type Skips string
+type SkipRange string
+
+type BundleObject struct {
+	File `json:",inline"`
+}
+
+type File struct {
+	ref  string
+	data []byte
+}
+
+type fileJSON struct {
+	Ref  string `json:"ref,omitempty"`
+	Data []byte `json:"data,omitempty"`
+}
+
+func (f *File) UnmarshalJSON(data []byte) error {
+	var t fileJSON
+	if err := json.Unmarshal(data, &t); err != nil {
+		return err
+	}
+	if len(t.Ref) > 0 && len(t.Data) > 0 {
+		return errors.New("fields 'ref' and 'data' are mutually exclusive")
+	}
+	f.ref = t.Ref
+	f.data = t.Data
+	return nil
+}
+
+func (f File) MarshalJSON() ([]byte, error) {
+	return json.Marshal(fileJSON{
+		Ref:  f.ref,
+		Data: f.data,
+	})
+}
+
+func (f File) IsRef() bool {
+	return len(f.ref) > 0
+}
+
+func (f File) GetRef() string {
+	return f.ref
+}
+
+func (f File) GetData(root, cwd string) ([]byte, error) {
+	if !f.IsRef() {
+		return f.data, nil
+	}
+	rootAbs, err := filepath.Abs(root)
+	if err != nil {
+		return nil, err
+	}
+	if filepath.IsAbs(f.ref) {
+		return nil, fmt.Errorf("reference must be a relative path")
+	}
+	refAbs, err := filepath.Abs(filepath.Join(cwd, f.ref))
+	if err != nil {
+		return nil, err
+	}
+	if !strings.HasPrefix(refAbs, rootAbs) {
+		return nil, fmt.Errorf("reference %q must be within root %q", refAbs, rootAbs)
+	}
+	return ioutil.ReadFile(refAbs)
+}
+
+type Properties struct {
+	Packages         []Package
+	PackagesRequired []PackageRequired
+	Channels         []Channel
+	GVKs             []GVK
+	GVKsRequired     []GVKRequired
+	Skips            []Skips
+	SkipRanges       []SkipRange
+	BundleObjects    []BundleObject
+
+	Others []Property
+}
+
+const (
+	TypePackage         = "olm.package"
+	TypePackageRequired = "olm.package.required"
+	TypeChannel         = "olm.channel"
+	TypeGVK             = "olm.gvk"
+	TypeGVKRequired     = "olm.gvk.required"
+	TypeSkips           = "olm.skips"
+	TypeSkipRange       = "olm.skipRange"
+	TypeBundleObject    = "olm.bundle.object"
+)
+
+func Parse(in []Property) (*Properties, error) {
```

Comment:
This may be a timing issue again, but I would expect this to use the pluggable property validation that vu added in `api`